docs/userguide/src/portingguide/concerns/weakref.md

```diff
@@ -0,0 +1,477 @@
+            shall hold a valid reference to the referent before it is cleared.)
+    -   If the referent has not been reached, yet, then
+        -   clear the referent field,
+        -   remove the `PhantomReference` from the list of phantom references, and
+        -   optionally enqueue it to the associated `ReferenceQueue` if it has one.
+    -   (This step cannot expand the transitive closure.)
+
+As an optimization,
+
+-   Step 1 can be, as we described in the previous section, eliminated by merging it with the strong
+    closure in non-emergency GC, or with `WeakReference` processing in emergency GC.
+-   Step 2 can be merged with Step 3 since Step 2 never expands the transitive closure.
+
+Therefore, we only need to run `process_weak_refs` twice:
+
+1.  Handle `WeakReference` (and also `SoftReference` in emergency GC), and then handle finalizable
+    objects.
+2.  Handle `PhandomReference`.
+
+To implement this, the VM binding may need to implement some kind of *state machine* so that the
+`Scanning::process_weak_refs` function behaves differently each time it is called.  For example,
+
+```rust
+fn process_weak_refs(...) -> bool {
+    let mut state = /* Get VM-specific states here. */;
+
+    match *state {
+        State::ProcessSoftReference => {
+            process_soft_references(...);
+            *state = State::ProcessWeakReference;
+            return true; // Run this function again.
+        }
+        State::ProcessWeakReference => {
+            process_weak_references(...);
+            *state = State::ProcessFinalizableObjects;
+            return true; // Run this function again.
+        }
+        State::ProcessFinalizableObjects => {
+            process_finalizable_objects(...);
+            *state = State::ProcessPhantomReferences;
+            return true; // Run this function again.
+        }
+        State::ProcessPhantomReferences => {
+            process_phantom_references(...);
+            *state = State::ProcessSoftReference
+            return false; // Proceed to the Release stage.
+        }
+    }
+}
+```
+
+### Ephemerons
+
+An [Ephemeron] has a *key* and a *value*, both of which are object references.  The key is a weak
+reference, while the value keeps the referent alive only if both the ephemeron itself and the key
+are reachable.
+
+[Ephemeron]: https://dl.acm.org/doi/10.1145/263700.263733
+
+To support ephemerons, the VM binding needs to identify ephemerons.  This includes ephemerons as
+individual objects, objects that contain ephemerons, and, equivalently, objects that contain
+key/value fields that have semantics similar to ephemerons.
+
+The following is the algorithm for processing ephemerons.  It gradually discovers ephemerons as we
+do the tracing.  We maintain a queue of ephemerons which is empty before the `Closure` stage.
+
+1.  In `Scanning::scan_object` and `Scanning::scan_object_and_trace_edges`, we enqueue ephemerons as
```

Comment:
Have to phrase this a bit delicately since there are two queues here. The ephemeron queue and the scanning/tracing queue. I got tripped up by you saying "enqueuing ephemerons but not tracing them".

Reply:
Yes.  This "enqueue" refers to the ephemeron queue.